source/trigger/cdc.go

```diff
@@ -0,0 +1,353 @@
+// Copyright © 2022 Meroxa, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package trigger
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	sq "github.com/Masterminds/squirrel"
+	"github.com/conduitio/conduit-connector-postgres/common"
+	sdk "github.com/conduitio/conduit-connector-sdk"
+	"github.com/jackc/pgx/v4"
+	"github.com/jackc/pgx/v4/pgxpool"
+	"go.uber.org/multierr"
+)
+
+const (
+	timeoutBeforeCloseDBSec        = 20
+	timeoutToClearTrackingTableSec = 5
+)
+
+var errWrongTrackingOperationType = errors.New("wrong tracking operation type")
+
+// CDC is an implementation of a CDC iterator in a trigger CDC mode.
+type CDC struct {
+	conn     *pgxpool.Pool
+	position *Position
+	rows     pgx.Rows
+
+	// tableSrv service for clearing tracking table.
+	tableSrv *trackingTableService
+
+	// table is a table name.
+	table string
+	// orderingColumn is a name of column what iterator use for sorting data.
+	orderingColumn string
+	// key is a column's name that iterator will use for setting key in record.
+	key string
+	// columns is a list of table's columns for record payload.
+	// If empty - will get all columns.
+	columns []string
+	// batchSize is a size of batch.
+	batchSize uint64
+	// conduit is names of conduit items in DB like tracking table, trigger, and procedure.
+	conduit string
+}
+
+// CDCParams is an incoming params for the NewCDC function.
+type CDCParams struct {
+	Conn           *pgxpool.Pool
+	Position       *Position
+	Table          string
+	OrderingColumn string
+	Key            string
+	Columns        []string
+	BatchSize      uint64
+	Conduit        string
+}
+
+type trackingTableService struct {
+	m sync.Mutex
+
+	// channel for the stop signal
+	stopCh chan struct{}
+	// channel to notify that all queries are completed and the database connection can be closed
+	canCloseCh chan struct{}
+	// error channel
+	errCh chan error
+	// slice of identifiers to delete
+	idsToDelete []any
+}
+
+// NewCDC creates a new instance of the CDC iterator.
+func NewCDC(ctx context.Context, params CDCParams) (*CDC, error) {
+	var err error
```

Comment:
Do we actually need this error to be declared here?

Reply:
No, of course not, it has to be deleted.
Done.